### PR TITLE
Issue151 monetdbe bind

### PR DESCRIPTION
**Pull request by aris-koning**
This pull request should accommodate the recently introduced API changes in monetdbe_prepare. First is to make sure that there are no regressions. Hopefully the CI can confirm that this is the current situation.

While developing my API accommodations I was hitting problems in the connection boilerplate causing regressions on my own system while running the unit tests. I think I have fixed what needed to be fixed there. Ideally this should have been fixed in a separate PR. But life is hard sometimes. The unit tests are accordingly fixed/restructured.

Next step is to actually make sensible use of the new API changes. But again first lets see how the current PR survives the CI.

### CODE REVIEW

**Comment by gijzelaerr** [comment]
awesome, green! thanks, will have a look

File: .github/workflows/linux.yml [hunk 1]
```diff
@@ -6,6 +6,8 @@ on:
     tags: [ "*" ]
   pull_request:
     branches: [ master ]
+  # Allows you to run this workflow manually from the Actions tab
+  workflow_dispatch:
```
Comment (gijzelaerr) [inline .github/workflows/linux.yml:10]: thats new to me, how does this work?

File: .github/workflows/osx.yml [hunk 2]
```diff
@@ -15,46 +17,32 @@ jobs:
     steps:
       -
         name: brew packages
-        run: brew install monetdb
-      - 
+        run: brew install bison
```
Comment (gijzelaerr) [inline .github/workflows/osx.yml:20]: why build Jul2021? there is a binary bottle installable with homebrew, which makes the CI complete quicker.

File: monetdbe/_cffi/convert/bind.py [hunk 3]
```diff
@@ -10,8 +10,8 @@ def monetdbe_int(data: int) -> ffi.CData:
         return ffi.new("int *", data)
 
 
-def bind_str(data: str) -> bytes:
-    return str(data).encode()
+def bind_str(data: str) -> ffi.CData:
+    return ffi.new("char[]", str(data).encode())
```
Comment (gijzelaerr) [inline monetdbe/_cffi/convert/bind.py:14]: why is this required?

File: monetdbe/_cffi/internal.py [hunk 4]
```diff
@@ -180,11 +169,13 @@ def open(self) -> monetdbe_database:
                 lib.monetdbe_close(connection)
             else:
                 error = errors.get(result_code, "unknown error")
-            raise exceptions.OperationalError(f"Failed to open database: {error} (code {result_code})")
+            msg = "Failed to open database: {error} (code {result_code})".format(error=error, result_code=result_code)
```
Comment (gijzelaerr) [inline monetdbe/_cffi/internal.py:172]: the `.format()` is quite old and not really the preferred string formatting method anymore for py>3.5

File: monetdbe/_cffi/internal.py [hunk 5]
```diff
@@ -268,7 +257,7 @@ def append(self, table: str, data: Mapping[str, np.ndarray], schema: str = 'sys'
     def prepare(self, query: str) -> monetdbe_statement:
         self._switch()
         stmt = ffi.new("monetdbe_statement **")
-        check_error(lib.monetdbe_prepare(self._monetdbe_database, str(query).encode(), stmt))
+        check_error(lib.monetdbe_prepare(self._monetdbe_database, str(query).encode(), stmt, ffi.NULL))
```
Comment (gijzelaerr) [inline monetdbe/_cffi/internal.py:260]: how does this work with Jul2021?

File: monetdbe/_cffi/internal.py [hunk 6]
```diff
@@ -296,3 +285,14 @@ def get_columns(self, table: str, schema: str = 'sys') -> Iterator[Tuple[str, in
...
             type_ = types_p[0][i]
             yield name, type_
+
+
+from monetdbe._cffi.branch import newer_then_jul2021
+if not newer_then_jul2021:
+    def prepare(self, query: str) -> monetdbe_statement:
+        self._switch()
+        stmt = ffi.new("monetdbe_statement **")
+        check_error(lib.monetdbe_prepare(self._monetdbe_database, str(query).encode(), stmt))
```
Comment (gijzelaerr) [inline monetdbe/_cffi/internal.py:295]: ah now i get it, you overwrite it dynamically. I'm surprised mypy doesn't complain about the signature here.

Maybe it is prettier and less confusing to do something like:

```
class MonetThing:
   def _prepare_jul2021(self, *, **):
       ...

   def _prepare_default(self, *, **):
       ...

   def prepare(self, *args, **kwargs):
        if jul2021:
            self._prepare_jul2021(*args, **kwargs)
        else:
            self._prepare_default(*args, **kwargs)
```

File: monetdbe/_cffi/internal.py [hunk 7]
```diff
@@ -296,3 +285,14 @@ def get_columns(self, table: str, schema: str = 'sys') -> Iterator[Tuple[str, in
...
+        return stmt[0]
+
+    setattr(Internal, 'prepare', prepare)
```
Comment (gijzelaerr) [inline monetdbe/_cffi/internal.py:298]: ah yes, see :) 

in case you want to silent mypy it is also okay to add `# type: ignore`, but then you need to have a good reason.

File: monetdbe/cursors.py [hunk 8]
```diff
@@ -37,6 +37,12 @@ def __init__(self, con: 'Connection'):
 
         self._fetch_generator: Optional[Iterator['Row']] = None
 
+    def __enter__(self):
```
Comment (gijzelaerr) [inline monetdbe/cursors.py:40]: nice, don't know why i never added this.

File: tests/test_cffi.py [hunk 9]
```diff
@@ -15,81 +15,81 @@ def test_cffi(self):
...
-        con = connect()
-        con.execute("CREATE TABLE test (i int)")
-        data = {'i': np.array([1, 2, 3]), 'j': np.array([1, 2, 3])}
-        with self.assertRaises(ProgrammingError):
-            con._internal.append(table='test', data=data)
+        with connect() as con:
+            con.execute("CREATE TABLE test (i int)")
+            data = {'i': np.array([1, 2, 3]), 'j': np.array([1, 2, 3])}
+            with self.assertRaises(ProgrammingError):
+                con._internal.append(table='test', data=data)
```
Comment (gijzelaerr) [inline tests/test_cffi.py:22]: this all comes from the sqlite test suite, which assumes the connections are closed by the garbage collector at some point.

but yes it is much better be explicit and use a context manager

File: tests/test_dbapi.py [hunk 10]
```diff
@@ -122,11 +122,6 @@ def test_RollbackAfterNoChanges(self):
     def test_Cursor(self):
         cu = self.cx.cursor()
 
-    def test_FailedOpen(self):
```
Comment (gijzelaerr) [inline tests/test_dbapi.py:125]: why remove?

File: tests/test_dbapi.py [hunk 11]
```diff
@@ -143,28 +138,23 @@ def test_Exceptions(self):
         self.assertEqual(self.cx.ProgrammingError, monetdbe.ProgrammingError)
         self.assertEqual(self.cx.NotSupportedError, monetdbe.NotSupportedError)
 
-    def test_InTransaction(self):
```
Comment (gijzelaerr) [inline tests/test_dbapi.py:146]: also here why remove?

File: tests/test_dbapi.py [hunk 12]
```diff
@@ -178,16 +168,48 @@ def __fspath__(self):
...
 
-    @unittest.skip("TODO: Not yet implemented, see issue #22")
-    def test_OpenUri(self):
-        self.addCleanup(rmtree, TESTFN)
-        with monetdbe.connect(TESTFN) as cx:
-            cx.execute('create table test(id integer)')
-        with monetdbe.connect('file:' + TESTFN, uri=True) as cx:
-            cx.execute('insert into test(id) values(0)')
-        with monetdbe.connect('file:' + TESTFN + '?mode=ro', uri=True) as cx:
+    def test_InTransaction(self):
```
Comment (gijzelaerr) [inline tests/test_dbapi.py:171]: looks like you changed the order of these tests which messes with the github pr diff.

File: tests/test_lite/test_dbapi03.py [hunk 13]
```diff
@@ -5,48 +5,52 @@
 
 class TestShutdown:
     def test_commited_on_restart(self, monetdbe_cursor_autocommit):
-        (cursor, connection, dbfarm) = monetdbe_cursor_autocommit
+        context = monetdbe_cursor_autocommit
```
Comment (gijzelaerr) [inline tests/test_lite/test_dbapi03.py:8]: these test suites are a bit messy if i remember correctly, sorry about that. I just copied them from the monetdb lite repo and modified them to fit our less-globals-api.

File: tests/test_regression.py [hunk 14]
```diff
@@ -295,7 +299,7 @@ def collation_cb(a, b):
 
     def test_RecursiveCursorUse(self):
         # note: modified test slighty since we actually just handle this fine.
-        con = monetdbe.connect(":memory:")
+        con = self.con
```
Comment (gijzelaerr) [inline tests/test_regression.py:302]: why is this required?

File: .github/workflows/linux.yml (hunk 1 above)
Comment (gijzelaerr) [reply .github/workflows/linux.yml:10]: I missed the conversation on the slack channel, I understand now.

File: .github/workflows/osx.yml (hunk 2 above)
Comment (gijzelaerr) [reply .github/workflows/osx.yml:20]: I missed the conversation on the slack channel, I understand now.

File: monetdbe/_cffi/internal.py (hunk 5 above)
Comment (aris-koning) [reply monetdbe/_cffi/internal.py:260]: On Jul2021 there is a method override for the `prepare` at the end of internal.py. It is a bit ugly but the most concise and direct approach.

See

https://github.com/MonetDBSolutions/MonetDBe-Python/blob/16561d947406c01c6a9bf51f89e6b9540ed27796/monetdbe/_cffi/internal.py#L290-L298

File: monetdbe/_cffi/internal.py (hunk 6 above)
Comment (aris-koning) [reply monetdbe/_cffi/internal.py:295]: In my case it happens at static runtime. While in your case it happens every time prepare is called. A small performance consideration basically.

File: monetdbe/_cffi/internal.py (hunk 7 above)
Comment (aris-koning) [reply monetdbe/_cffi/internal.py:298]: indeed

Comment (aris-koning) [reply monetdbe/_cffi/internal.py:298]: I'll keep that in mind

File: monetdbe/_cffi/internal.py (hunk 6 above)
Comment (gijzelaerr) [reply monetdbe/_cffi/internal.py:295]: true, but if performance is of the highest importance people probably would not use python anyway. I would not sacrifice type consistency for one reasonably high-level function call less.

Comment (aris-koning) [reply monetdbe/_cffi/internal.py:295]: We have the CI on default and Juli2021 to make sure that everything is working correctly. I could do it more OOP/Pythonic by using base and derived classes to do the override and maintain the 'performance'. But the reason why I prefer not to is because this  is basically a temporal compact code patch and that I **want** it to be ugly, concise and readily removed which will happen once Jul2021 is old enough. This discussion is a bit subjective I admit but I prefer to keep it.

File: tests/test_regression.py (hunk 14 above)
Comment (aris-koning) [reply tests/test_regression.py:302]: In the setUp a connection object is already created. There is no reason for this test not to use that. And the previous code was either leaking connection objects or it trusted on the GC to timely destroy the connection object. Unless you take explicit control over the GC, this can become problematic once you try to open a connection to a _different_ db because monetdbe does not currently allow concurrent connections to different db's.

Comment (gijzelaerr) [reply tests/test_regression.py:302]: ok good, just curious what your motivations were here.

File: monetdbe/_cffi/internal.py (hunk 6 above)
Comment (gijzelaerr) [reply monetdbe/_cffi/internal.py:295]: I guess this is more a matter of style, and I gave you my pythonic style view on the matter. I actually quite dislike OO also, the only reason I use it in the API is that the DBAPI dictates it. I don't want to push you to implement things in a certain way and I trust your judgment, so if you think this is fine then lets just do it this way.

File: tests/test_dbapi.py (hunk 11 above)
Comment (aris-koning) [reply tests/test_dbapi.py:146]: It is still there but I moved it to `VariousConnectionTests` because there were some problems with the Connection object management.

File: tests/test_dbapi.py (hunk 10 above)
Comment (aris-koning) [reply tests/test_dbapi.py:125]: Forgive me. I actually just moved `test_FailedOpen` to `VariousConnectionTests` because of Connection object management.

File: monetdbe/_cffi/internal.py (hunk 4 above)
Comment (aris-koning) [reply monetdbe/_cffi/internal.py:172]: 👍 will be in the next push.

File: monetdbe/_cffi/convert/bind.py (hunk 3 above)
Comment (aris-koning) [reply monetdbe/_cffi/convert/bind.py:14]: Hmm I think because monetdbe expects a deep copy (so an array) of the C string and not a reference to the original. I remembered that I got corruption errors in certain situations without this change.
